File: SerializeDeserializeTree.java
// Time Complexity : O(n) serialize, O(n) deserialize
// Space Complexity : O(n), for both queue + arraylist 
// Did this code successfully run on Leetcode : 10/50 passed, bug in deserialization code

// Your code here along with comments explaining your approach
// temp work, will update after class

public class Codec {
    
    int maxDepth = 0;
    
    private void helper(TreeNode node, int curDepth){
        if(node==null)
            return;
        
        helper(node.left, curDepth+1);
        if(curDepth > maxDepth)
            maxDepth = curDepth;        
        helper(node.right, curDepth+1);
    }
    
    // Encodes a tree to a single string.
    public String serialize(TreeNode root) {
        if(root==null)
            return "";
        
        maxDepth = 1;
        helper(root, 1);
        
        StringBuilder sb = new StringBuilder();
        sb.append("[");
        
        Queue<TreeNode> queue = new LinkedList<TreeNode>();
        queue.add(root);
        int level = 1;

        while(!queue.isEmpty()){
            int size = queue.size();

            for(int i=0; i<size; i++){
                TreeNode node = queue.poll();
                if(node==null)
                    sb.append("null");
                else
                    sb.append(Integer.toString(node.val));
                
                if(node!=null && level<maxDepth){
                    queue.offer(node.left);
                    queue.offer(node.right);                    
                }
                
                sb.append(",");
            }
            
            level++;
        }
        if(sb.length()>1){
            sb.deleteCharAt(sb.length()-1);
        }
        sb.append("]");
        
        return sb.toString();
    }

    // Decodes your encoded data to tree.
    public TreeNode deserialize(String data) {
        if(data==null || data.length()<3)
            return null;
        
        String temp = data.substring(1, data.length()-1);
        String[] parts = temp.split(",");
        
        List<TreeNode> listOfNodes = new ArrayList<>();
        for(int i=0; i<parts.length; i++){
            if(parts[i].equals("null")){
                listOfNodes.add(null);
            }
            else{
                int val = Integer.parseInt(parts[i]);
                listOfNodes.add(new TreeNode(val));    
            }
        }
        
        TreeNode root = listOfNodes.get(0);
        int n = listOfNodes.size();
        for(int i=0; i<=(n/2)-1; i++){
            TreeNode node = listOfNodes.get(i);
            
            if(node!=null){
                node.left = listOfNodes.get((2*i)+1);    
                node.right = listOfNodes.get((2*i)+2);    
            }
        }
        
        return root;
    }
}

File: serialize_deserialize_tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


# Time - O(N)
# Space - O(N)

class Codec:

    def serialize(self, root):
        """Encodes a tree to a single string.
        
        :type root: TreeNode
        :rtype: str
        """
        result = []
        
        if not root:
            return result
        
        q = deque()
        q.append(root)
        
        while q:
            
            current_node = q.popleft()
            
            if current_node:
                # if not a null node
                result.append(str(current_node.val))
                # result.append(',') # not needed for list but string builder
                
                # add left right children to queue not result, no need to check for null
                q.append(current_node.left)
                q.append(current_node.right)
                
            else:
                # if null node, simply add to result, do not check for children
                result.append('null')
                # result.append(',')
                
        return result
        

    def deserialize(self, data):
        """Decodes your encoded data to tree.
        
        :type data: str
        :rtype: TreeNode
        """
        # result string from above is fed to this function
        
        if not data:
            return None
        
        # data.split(',')
        q = deque()
        i = 0
        
        # construct root node with 0th index val from data
        root_node = TreeNode(int(data[i]))
        q.append(root_node)
        
        # increment i to point next index which is left child of root node
        i += 1
        
        # as long as q is not empty and i does not breach
        while q and i < len(data):
            current_node = q.popleft()
            
            # now construct left child of current node (i + 1 position)
            if data[i] != 'null':
                current_node.left = TreeNode(int(data[i]))
                q.append(current_node.left)
                
            i += 1 # whether it was null or not inc i to next -- so that when we pop next we can use this location node as children
            
            # now construct right child (ie at i + 2 position)
            
            if data[i] != 'null':
                current_node.right = TreeNode(int(data[i]))
                q.append(current_node.right)
                
            i += 1 # similar to above inc it to next one -- so that when we pop next we can use this location node as children
        
        
        return root_node

# Your Codec object will be instantiated and called as such:
# ser = Codec()
# deser = Codec()
# ans = deser.deserialize(ser.serialize(root))